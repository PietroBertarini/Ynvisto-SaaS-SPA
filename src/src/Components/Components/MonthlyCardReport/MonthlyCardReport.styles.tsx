import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

const styledBy = (property, mapping) => (props) => mapping[props[property]];
export const useStyles = makeStyles((theme: Theme) => createStyles({
  titleContainer: {
    width: '100%',
    height: '4vw',
    paddingTop: '0!important',
    paddingBottom: '0vw!important',
    paddingLeft: '0!important',
    paddingRight: '0!important',
    background: '#28282F',
  },
  card: {
    width: '100%',
    height: '100%',
    borderRadius: 20,
    background: 'transparent',
  },
  content: {
    width: '100%',
    height: '20vw',
    minHeight: 300,
    background: '#1E1E23',
  },
  title: {
    color: '#FFFFFF',
    fontFamily: 'Zurich',
    fontWeight: 'bold',
  },
  actionArea: {
    display: 'inline-block',
  },
  titleContainerBox: {
    width: '100%',
    height: '100%',
  },
  headerMonetaryValue: {
    textAlign: 'center',
    color: styledBy('status', {
      GOOD: '#05E985',
      WARNING: '#E9C46A',
      BAD: '#EB5050',
    }),
    marginBottom: '1vw',
    fontWeight: 'bold',
  },
  modal: {
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    width: '100vw',
    '& .MuiDialog-container': {
      width: '100%',
    },
  },
  paper: {
    backgroundColor: '#1E1E23',
    boxShadow: theme.shadows[5],
    padding: theme.spacing(2, 4, 3),
    width: '25vw',
  },
  showDiv: {
    width: '90%',
    color: '#ccc',
    borderTop: '1px solid #ccc',
    padding: '20',
    position: 'absolute',
    bottom: '0',
  },
  showMore: {
    padding: '10',
    textAlign: 'center',
  },
}));
