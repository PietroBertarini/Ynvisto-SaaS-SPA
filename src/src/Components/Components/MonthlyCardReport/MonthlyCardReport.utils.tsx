import '../../../Theme/sweetalert2/sweetalert2.scss';
import { allTabs } from '../../Pages/TaxReport/TaxReport.utils';
import MonthlyTaxReport from '../../../Classes/MonthlyTaxReport/MonthlyTaxReport.class';
import { ITaxReport } from '../../../Redux/Services/TaxReports/TaxReports.actions';

export interface ITaxdDetailInfo {
  commonOps?: {
    totalSell: number,
    monthProfit: number,
    taxableProfit: number,
    compensatedLoss: number,
    calcBase: number,
    taxDue: number,
  },
  dayTrade?: {
    taxableProfit: number,
    compensatedLoss: number,
    calcBase: number,
    taxDue: number,
  },
  fii?: {
    liquidResult: number,
    compensatedLoss: number,
    calcBase: number,
    taxDue: number,
  },
  finalResults: {
    commonOpTax: number,
    daytradeTax: number,
    withholdingTax: number,
    totalTax: number,
    fiiTax: number,
    previousDarf: number,
    payableTax: number,
  },
}

const defaultTaxdDetailInfo: ITaxdDetailInfo = {
  finalResults: {
    commonOpTax: 0,
    daytradeTax: 0,
    withholdingTax: 0,
    totalTax: 0,
    fiiTax: 0,
    previousDarf: 0,
    payableTax: 0,
  },
};

export const getMonthNumber = (monthName: string) : number => {
  const index = allTabs.findIndex((element) => element === monthName);
  return index === -1 ? 0 : (index);
};

export function getTaxDetailsInfo(monthlyTaxReport ?: MonthlyTaxReport) : ITaxdDetailInfo {
  if (monthlyTaxReport) {
    const commonOpsValidate = monthlyTaxReport.commonOpsReport
      && ((monthlyTaxReport.commonOpsReport.totalSaleVolume !== 0)
      || (monthlyTaxReport.commonOpsReport.profits !== 0)
      || (monthlyTaxReport.commonOpsReport.taxableProfits !== 0)
      || (monthlyTaxReport.commonOpsReport.previousMonthLosses !== 0)
      || (monthlyTaxReport.commonOpsReport.taxBase !== 0)
      || (monthlyTaxReport.commonOpsReport.taxOwed !== 0));

    const commonOps = monthlyTaxReport.commonOpsReport && commonOpsValidate ? {
      totalSell: monthlyTaxReport.commonOpsReport.totalSaleVolume,
      monthProfit: monthlyTaxReport.commonOpsReport.profits,
      taxableProfit: monthlyTaxReport.commonOpsReport.taxableProfits,
      compensatedLoss: monthlyTaxReport.commonOpsReport.previousMonthLosses,
      calcBase: monthlyTaxReport.commonOpsReport.taxBase,
      taxDue: monthlyTaxReport.commonOpsReport.taxOwed,
    } : undefined;

    const dayTradeValidate = monthlyTaxReport.dayTradeReport
      && ((monthlyTaxReport.dayTradeReport.profits !== 0)
      || (monthlyTaxReport.dayTradeReport.previousMonthLosses !== 0)
      || (monthlyTaxReport.dayTradeReport.taxBase !== 0)
      || (monthlyTaxReport.dayTradeReport.taxesOwed !== 0));

    const dayTrade = monthlyTaxReport.dayTradeReport && dayTradeValidate ? {
      taxableProfit: monthlyTaxReport.dayTradeReport.profits,
      compensatedLoss: monthlyTaxReport.dayTradeReport.previousMonthLosses,
      calcBase: monthlyTaxReport.dayTradeReport.taxBase,
      taxDue: monthlyTaxReport.dayTradeReport.taxesOwed,
    } : undefined;

    const fiiValidate = monthlyTaxReport.fiiReport
      && ((monthlyTaxReport.fiiReport.profits !== 0)
      || (monthlyTaxReport.fiiReport.previousMonthLosses !== 0)
      || (monthlyTaxReport.fiiReport.taxBase !== 0)
      || (monthlyTaxReport.fiiReport.netTaxes !== 0));

    const fii = monthlyTaxReport.fiiReport && fiiValidate ? {
      liquidResult: monthlyTaxReport.fiiReport.profits,
      compensatedLoss: monthlyTaxReport.fiiReport.previousMonthLosses,
      calcBase: monthlyTaxReport.fiiReport.taxBase,
      taxDue: monthlyTaxReport.fiiReport.netTaxes,
    } : undefined;

    const finalResults = {
      commonOpTax: monthlyTaxReport.commonOpsReport?.taxOwed || 0,
      daytradeTax: monthlyTaxReport.dayTradeReport?.taxesOwed || 0,

      withholdingTax: (monthlyTaxReport.commonOpsReport?.irrfToCompensate || 0)
        + (monthlyTaxReport.dayTradeReport?.irrfToCompensate || 0),

      totalTax: monthlyTaxReport.stockMarketNetTaxes,
      fiiTax: monthlyTaxReport.fiiReport?.netTaxes || 0,
      previousDarf: monthlyTaxReport.previousDarf,

      payableTax: (monthlyTaxReport.fiiReport?.netTaxes || 0)
        + monthlyTaxReport.stockMarketNetTaxes
        + monthlyTaxReport.previousDarf,
    };

    return {
      commonOps,
      dayTrade,
      fii,
      finalResults,
    };
  }

  return defaultTaxdDetailInfo;
}

/**
 * Search at TaxReport redux state if contains the report of this year of input monthNumber
 * @param taxReport: Tax Report Redux State from API
 * @param monthNumber: Month number to get the report (January:1, Feb: 2,..., December: 12)
 */
export const getMonthReport = (monthNumber: number, taxReport?: ITaxReport)
    : MonthlyTaxReport | undefined => {
  const thisYear = new Date().getFullYear();
  if (taxReport) {
    return taxReport.monthlyReports.find((item) => (item.monthIndex === monthNumber - 1)
        && item.year === thisYear);
  }
  return undefined;
};
