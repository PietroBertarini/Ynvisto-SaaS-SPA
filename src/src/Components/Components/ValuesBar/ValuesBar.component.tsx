import React from 'react';
import { ResponsiveBar } from '@nivo/bar';
import { Card } from '@material-ui/core';
import { useStyles } from './ValuesBar.styles';
import { getChartData } from './ValuesBar.utils';
import { IMonthlyCardData } from '../MonthlyCardReportLegend/MonthlyCardReportLegend.types';
import { getTotalValue } from '../MonthlyCardReportLegend/MonthlyCardReportLegend.utils';

export interface OwnProps {
    /**
     *  The data info to construct the legend and the bar of the card.
     */
    data : IMonthlyCardData[];
}

/**
 * Show a progress bar using multiple colors and divs.
 */
function ValuesBar(props: OwnProps) {
  const classes = useStyles();
  const { data } = props;
  const totalVolume = getTotalValue(data);

  const chartData = getChartData(data, totalVolume);
  const barData = chartData.valuesObj;
  const barColors = chartData.colors;
  const barKeys = chartData.keys;

  return (
    <Card variant="outlined" className={classes.card}>
      <ResponsiveBar
        data={[barData]}
        keys={barKeys}
        layout="horizontal"
        colors={barColors}
        borderWidth={1}
        borderColor={{ from: 'color', modifiers: [['darker', 1.6]] }}
        axisTop={null}
        axisRight={null}
        axisBottom={null}
        axisLeft={null}
        enableGridY={false}
        enableLabel={false}
        animate
        isInteractive={false}
        motionStiffness={90}
        motionDamping={15}
      />
    </Card>
  );
}

export default ValuesBar;
