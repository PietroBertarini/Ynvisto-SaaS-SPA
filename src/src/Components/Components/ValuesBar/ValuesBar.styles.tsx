import { makeStyles, createStyles } from '@material-ui/core/styles';

export const useStyles = makeStyles(() => createStyles({
  card: {
    height: 30,
    marginBottom: '1.5vw',
    borderRadius: 50,
    backgroundColor: 'transparent',
    '& > div': {
      width: '100%!important',
      height: '200%!important',
      marginTop: '-10px',
    },
  },
}));
