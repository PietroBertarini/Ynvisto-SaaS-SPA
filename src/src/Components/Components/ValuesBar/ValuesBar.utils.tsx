import { IMonthlyCardData } from '../MonthlyCardReportLegend/MonthlyCardReportLegend.types';
import { isSameSignAndNotZero } from '../../../Util/general.utils';

export interface ChartData {
  valuesObj: Record<string, number>;
  colors: string[];
  keys: string[];
}

export function getChartData(data :IMonthlyCardData[], headerMonetaryValue : number) : ChartData {
  const object: Record<string, number> = {};
  const isPositiveHeader = headerMonetaryValue > 0;

  const haveSomeValueAtBat = isPositiveHeader
    ? data.find((item) => item.value > 0)
    : data.find((item) => item.value < 0);

  if (!haveSomeValueAtBat) {
    return {
      valuesObj: {
        default: 1,
      },
      colors: ['#B1B1B1'],
      keys: ['default'],
    };
  }

  // Order the array by label in order for the object (that is auto-ordered by label) to match the colors
  // array
  data = data.sort((a, b) => a.label.localeCompare(b.label));

  const colorsArr: string[] = [];
  const allKeys : string[] = [];
  data.forEach((item) => {
    if (isSameSignAndNotZero(item.value, headerMonetaryValue)) {
      object[item.label] = item.value;
      colorsArr.push(item.color);
      allKeys.push(item.label);
    }
  });

  return {
    valuesObj: object,
    colors: colorsArr,
    keys: allKeys,
  };
}
