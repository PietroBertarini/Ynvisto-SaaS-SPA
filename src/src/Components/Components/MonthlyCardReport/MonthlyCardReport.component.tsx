import React, { useCallback, useState } from 'react';
import Typography from '@material-ui/core/Typography';
import {
  Card, CardActionArea,
} from '@material-ui/core';
import Box from '@material-ui/core/Box';
import Dialog from '@material-ui/core/Dialog';
import CardContent from '@material-ui/core/CardContent';
import { useStyles } from './MonthlyCardReport.styles';
import ValuesBar from '../ValuesBar/ValuesBar.component';
import MonthlyCardReportLegend from '../MonthlyCardReportLegend/MonthlyCardReportLegend.component';
import { IMonthlyCardData } from '../MonthlyCardReportLegend/MonthlyCardReportLegend.types';

export enum ETitleStatus {
  GOOD = 'GOOD',
  WARNING = 'WARNING',
  BAD = 'BAD',
}

export interface OwnProps {
  /**
   * Header Status. Directly controls the color of the header monetary value
   */
  status: ETitleStatus,
  /**
   * Title of the card
   */
  title: string,
  /**
   * Main monetary value displayed on at the top of the card
   */
  headerMonetaryValue: number,
  /**
   * The data info to construct the legend and the bar of the card.
   */
  data: IMonthlyCardData[],
  /**
   * Modal that will be shown if clicked.
   */
  detailsModal?: JSX.Element,
}

/**
 * Get the stores states that TaxComponents uses and observe.
 * @param state: the stored stated of the Ynvisto.
 */

/**
 * Card with % bar and legend to show some inf.
 */
function MonthlyCardReport(props : OwnProps) {
  const {
    title, headerMonetaryValue, data, detailsModal,
  } = props;
  const [open, setOpen] = useState(false);

  const handleOpen = () => {
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };

  // Empty function, according by ESlint documentation, to avoid empty arrow functions
  const emptyFunc = useCallback(() => {
    // do nothing
  }, []);

  const classes = useStyles(props);
  return (
    <>
      <Card className={classes.card} onClick={detailsModal ? handleOpen : emptyFunc}>
        <CardActionArea className={classes.actionArea}>
          <CardContent className={classes.titleContainer}>
            <Box
              display="flex"
              justifyContent="center"
              alignContent="center"
              alignItems="center"
              className={classes.titleContainer}
            >
              <Typography variant="h5" align="center" className={classes.title}>
                {title}
              </Typography>
            </Box>
          </CardContent>
          <CardContent className={classes.content}>
            <Box display="flex" flexDirection="column" alignSelf="center" justifyContent="center">
              <Typography variant="h4" className={classes.headerMonetaryValue}>
                {headerMonetaryValue.toLocaleString('pt-br', { style: 'currency', currency: 'BRL' })}
              </Typography>
              <ValuesBar
                data={data}
              />
              <MonthlyCardReportLegend legendData={data} />
            </Box>
            {detailsModal && (<div className={classes.showDiv}><p className={classes.showMore}>Mais Detalhes</p></div>)}
          </CardContent>
        </CardActionArea>
      </Card>

      <Dialog
        className={classes.modal}
        onClose={handleClose}
        scroll="body"
        PaperProps={{
          style: { borderRadius: 20, backgroundColor: 'transparent' },
        }}
        aria-labelledby="simple-dialog-title"
        open={open}
      >
        <div className={classes.paper}>
          {detailsModal}
        </div>
      </Dialog>
    </>
  );
}

export default MonthlyCardReport;
