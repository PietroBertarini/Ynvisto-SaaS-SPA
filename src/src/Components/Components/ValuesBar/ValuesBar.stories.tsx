import React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from '@storybook/react/types-6-0';

import ValuesBar, { OwnProps } from './ValuesBar.component';
import {
  getProfitDataNegativeAndPositive,
  getProfitDataRemainsNegative,
} from '../../../Mocks/MonthlyCardReport.mock';

export default {
  title: 'Components/ValuesBar',
  component: ValuesBar,
} as Meta;

const Template: Story<OwnProps> = (args) => <ValuesBar {...args} />;

export const PositiveValues = Template.bind({});

PositiveValues.args = {
  data: getProfitDataNegativeAndPositive,
};

export const NegativeValues = Template.bind({});

NegativeValues.args = {
  data: getProfitDataRemainsNegative,
};
