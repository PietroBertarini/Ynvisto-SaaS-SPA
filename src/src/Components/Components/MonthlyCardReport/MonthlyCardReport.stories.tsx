import React from 'react';
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Meta, Story } from '@storybook/react/types-6-0';

import MonthlyCardReport, { ETitleStatus, OwnProps } from './MonthlyCardReport.component';
import { getProfitDataDefault } from '../../../Mocks/MonthlyCardReport.mock';

export default {
  title: 'Components/MonthlyCardReport',
  component: MonthlyCardReport,
} as Meta;

const Template: Story<OwnProps> = (args) => <MonthlyCardReport {...args} />;

export const Default = Template.bind({});
Default.args = {
  status: ETitleStatus.GOOD,
  title: 'stories',
  headerMonetaryValue: 10000,
  data: getProfitDataDefault,

};
