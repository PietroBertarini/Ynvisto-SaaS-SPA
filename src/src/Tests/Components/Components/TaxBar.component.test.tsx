import React from 'react';
import { shallow } from 'enzyme';
import ValuesBar from '../../../Components/Components/ValuesBar/ValuesBar.component';
import { getProfitDataNegativeAndPositive } from '../../../Mocks/MonthlyCardReport.mock';

test('Renders ValuesBar component and match to previously snapshot', () => {
  const component = shallow(<ValuesBar
    data={getProfitDataNegativeAndPositive}
  />);
  expect(component).toMatchSnapshot();
});
