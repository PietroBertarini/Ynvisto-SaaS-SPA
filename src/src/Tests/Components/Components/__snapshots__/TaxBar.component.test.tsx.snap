// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders ValuesBar component and match to previously snapshot 1`] = `
<WithStyles(ForwardRef(Card))
  className="makeStyles-card-1"
  variant="outlined"
>
  <ResponsiveBar
    animate={true}
    axisBottom={null}
    axisLeft={null}
    axisRight={null}
    axisTop={null}
    borderColor={
      Object {
        "from": "color",
        "modifiers": Array [
          Array [
            "darker",
            1.6,
          ],
        ],
      }
    }
    borderWidth={1}
    colors={
      Array [
        "#192d00",
      ]
    }
    data={
      Array [
        Object {
          "Subscrição": -100000,
        },
      ]
    }
    enableGridY={false}
    enableLabel={false}
    isInteractive={false}
    keys={
      Array [
        "Subscrição",
      ]
    }
    layout="horizontal"
    motionDamping={15}
    motionStiffness={90}
  />
</WithStyles(ForwardRef(Card))>
`;
